async_retry.go

```diff
@@ -0,0 +1,133 @@
+package asyncretry
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/avast/retry-go/v4"
+)
+
+type AsyncRetryFunc func(ctx context.Context) error
+
+type AsyncRetry interface {
+	// Do calls f and retry if necessary.
+	// In most cases, you should call Do in a new goroutine.
+	Do(ctx context.Context, f AsyncRetryFunc, opts ...Option) error
+	// Shutdown shutdowns gracefully
+	Shutdown(ctx context.Context) error
+}
+
+type asyncRetry struct {
+	// FIXME use RWMutex instead
+	mu           sync.Mutex // guards wg and shutdownChan
+	wg           sync.WaitGroup
+	shutdownChan chan struct{}
+}
+
+func NewAsyncRetry() AsyncRetry {
+	return &asyncRetry{
+		wg:           sync.WaitGroup{},
+		shutdownChan: make(chan struct{}),
+	}
+}
+
+var InShutdownErr = fmt.Errorf("AsyncRetry is in shutdown")
+
+func (a *asyncRetry) Do(ctx context.Context, f AsyncRetryFunc, opts ...Option) (retErr error) {
+	a.mu.Lock()
+	select {
+	case <-a.shutdownChan:
+		return InShutdownErr
+	default:
+	}
+	a.wg.Add(1)
```

Comment:
[question]

Do I need to include this code in the mutex lock?
I think `WaitGroup.Add()` is synchronization primitive.

https://github.com/golang/go/blob/release-branch.go1.17/src/sync/waitgroup.go#L64

Reply:
Yes! If `a.wg.Add` were not in lock, graceful shutdown would be broken as below:

1. `Do` checks that asyncRetry.is not in shutdown
2. Before 1 increments `wg`, scheduler stop 1's goroutine
3. Shutdown is called, `close(shudownChan)`
4. `a.wg.Wait()` is called in a new goroutine, and it is possible that call returns immediately when other tasks is not running
5. 1's goroutine restart

Reply:
I see!
Thanks for explaining.

Reply:
I think many people think as you did, so I add a comment:
https://github.com/Kyash/async-retry/pull/1/commits/99c6d70296613e9b1a01716acc3c1e354edc8df7

Reply:
Very nice comment!
Thanks!!

Reply:
nits:
I thought it would be nice to include this comment on the same line!

```suggestion
	a.wg.Add(1)  // notice that this line should be in lock so that shutdown would not go ahead
```